### PR TITLE
*: rewrite builtin function str_to_date

### PR DESCRIPTION
**Pull request by XuHuaiyu**
to #4080
PTAL @zz-jason @breeswish

### CODE REVIEW

File: plan/typeinfer_test.go [hunk 1]
```diff
@@ -81,23 +81,23 @@ func (s *testPlanSuite) TestInferType(c *C) {
...
+	//tests = append(tests, s.createTestCase4Aggregations()...)
+	//tests = append(tests, s.createTestCase4InfoFunc()...)
+	//tests = append(tests, s.createTestCase4EncryptionFuncs()...)
+	//tests = append(tests, s.createTestCase4CompareFuncs()...)
+	//tests = append(tests, s.createTestCase4Miscellaneous()...)
+	//tests = append(tests, s.createTestCase4OpFuncs()...)
+	//tests = append(tests, s.createTestCase4OtherFuncs()...)
 	tests = append(tests, s.createTestCase4TimeFuncs()...)
-	tests = append(tests, s.createTestCase4LikeFuncs()...)
+	//tests = append(tests, s.createTestCase4LikeFuncs()...)
```
Comment (zz-jason) [inline plan/typeinfer_test.go]: forget to release the commented tests ?

File: util/types/time.go [hunk 2]
```diff
@@ -2043,6 +2043,49 @@ var dateFormatParserTable = map[string]dateFormatParser{
 	// "%y": yearTwoDigits,         // Year, numeric (two digits)
 }
 
+// GetFormatType checks the type(Duration, Date or Datetime) of a format string.
+func GetFormatType(format string) (isDuration, isDate bool) {
+	durationTokens := map[string]struct{}{
```
Comment (zz-jason) [inline util/types/time.go]: can we put these two `map`s outside this function ?

File: expression/builtin_time.go [hunk 3]
```diff
@@ -1257,37 +1259,128 @@ type strToDateFunctionClass struct {
...
+	format, isNull, err := strArg.EvalString(nil, ctx.GetSessionVars().StmtCtx)
+	if err != nil || isNull {
+		return
+	}
+	isDuration, isDate := types.GetFormatType(format)
+	if isDuration && !isDate {
+		tp = mysql.TypeDuration
+	} else if !isDuration && isDate {
+		tp = mysql.TypeDate
+	}
```
Comment (zz-jason) [inline expression/builtin_time.go]: what if `!isDuration && !isDate` and `isDuration && isDate` ?

Comment (XuHuaiyu) [reply expression/builtin_time.go]: both return mysql.TypeDatetime

File: util/types/time.go (hunk 2 above)
Comment (XuHuaiyu) [reply util/types/time.go]: We only use these maps in this function,
I think we can move them outside if necessary later.

File: expression/builtin_time.go [hunk 4]
```diff
@@ -1257,37 +1259,128 @@ type strToDateFunctionClass struct {
...
+	baseTimeBuiltinFunc
 }
 
-// eval evals a builtinStrToDateSig.
-// See https://dev.mysql.com/doc/refman/5.5/en/date-and-time-functions.html#function_str-to-date
-func (b *builtinStrToDateSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
+func (b *builtinStrToDateDateSig) evalTime(row []types.Datum) (types.Time, bool, error) {
```
Comment (zz-jason) [inline expression/builtin_time.go]: add the comment back

**Comment by zz-jason** [comment]
rest LGTM

**Comment by jackysp** [comment]
LGTM

**Pull request merged**